File: Management/Models/CanvasModels/Assignments/CanvasAssignment.cs
using CanvasModel.Discussions;
using CanvasModel.Submissions;

namespace CanvasModel.Assignments;

public record CanvasAssignment
(
  [property: JsonPropertyName("id")]
  ulong Id,

  [property: JsonPropertyName("name")]
  string Name,

  [property: JsonPropertyName("description")]
  string Description,

  [property: JsonPropertyName("created_at")]
  DateTime CreatedAt,

  [property: JsonPropertyName("has_overrides")]
  bool HasOverrides,

  [property: JsonPropertyName("course_id")]
  ulong CourseId,

  [property: JsonPropertyName("html_url")]
  string HtmlUrl,

  [property: JsonPropertyName("submissions_download_url")]
  string SubmissionsDownloadUrl,

  [property: JsonPropertyName("assignment_group_id")]
  ulong AssignmentGroupId,

  [property: JsonPropertyName("due_date_required")]
  bool DueDateRequired,

  [property: JsonPropertyName("max_name_length")]
  uint MaxNameLength,

  [property: JsonPropertyName("peer_reviews")]
  bool PeerReviews,

  [property: JsonPropertyName("automatic_peer_reviews")]
  bool AutomaticPeerReviews,

  [property: JsonPropertyName("position")]
  ulong Position,

  [property: JsonPropertyName("grading_type")]
  string GradingType,

  [property: JsonPropertyName("published")]
  bool Published,

  [property: JsonPropertyName("unpublishable")]
  bool Unpublishable,

  [property: JsonPropertyName("only_visible_to_overrides")]
  bool OnlyVisibleToOverrides,

  [property: JsonPropertyName("locked_for_user")]
  bool LockedForUser,

  [property: JsonPropertyName("moderated_grading")]
  bool ModeratedGrading,

  [property: JsonPropertyName("grader_count")]
  uint GraderCount,

  [property: JsonPropertyName("allowed_attempts")]
  int AllowedAttempts,

  [property: JsonPropertyName("submission_types")]
  IEnumerable<string> SubmissionTypes,

  [property: JsonPropertyName("updated_at")]
  DateTime? UpdatedAt = null,

  [property: JsonPropertyName("due_at")]
  DateTime? DueAt = null,

  [property: JsonPropertyName("lock_at")]
  DateTime? LockAt = null,

  [property: JsonPropertyName("unlock_at")]
  DateTime? UnlockAt = null,

  [property: JsonPropertyName("all_dates")]
  IEnumerable<CanvasAssignmentDate>? AllDates = null,

  [property: JsonPropertyName("allowed_extensions")]
  IEnumerable<string>? AllowedExtensions = null,

  [property: JsonPropertyName("turnitin_enabled")]
  bool? TurnitinEnabled = null,

  [property: JsonPropertyName("vericite_enabled")]
  bool? VeriCiteEnabled = null,

  [property: JsonPropertyName("turnitin_settings")]
  CanvasTurnitinSettings? TurnitinSettings = null,

  [property: JsonPropertyName("grade_group_students_individually")]
  bool? GradeGroupStudentsIndividually = null,

  [property: JsonPropertyName("external_tool_tag_attributes")]
  CanvasExternalToolTagAttributes? ExternalToolTagAttributes = null,

  [property: JsonPropertyName("peer_review_count")]
  uint? PeerReviewCount = null,

  [property: JsonPropertyName("peer_reviews_assign_at")]
  DateTime? PeerReviewsAssignAt = null,

  [property: JsonPropertyName("intra_group_peer_reviews")]
  bool? IntraGroupPeerReviews = null,

  [property: JsonPropertyName("group_category_id")]
  ulong? GroupCategoryId = null,

  [property: JsonPropertyName("needs_grading_count")]
  uint? NeedsGradingCount = null,

  [property: JsonPropertyName("needs_grading_count_be_section")]
  IEnumerable<CanvasNeedsGradingCount>? NeedsGradingCountBySection = null,

  [property: JsonPropertyName("post_to_sis")]
  bool? PostToSis = null,

  [property: JsonPropertyName("integration_id")]
  string? IntegrationId = null,

  [property: JsonPropertyName("integration_data")]
  object? IntegrationData = null,

  [property: JsonPropertyName("muted")]
  bool? Muted = null,

  [property: JsonPropertyName("points_possible")]
  double? PointsPossible = null,

  [property: JsonPropertyName("has_submitted_submissions")]
  bool? HasSubmittedSubmissions = null,

  [property: JsonPropertyName("grading_standard_id")]
  ulong? GradingStandardId = null,

  [property: JsonPropertyName("lock_info")]
  CanvasLockInfo? LockInfo = null,

  [property: JsonPropertyName("lock_explanation")]
  string? LockExplanation = null,

  [property: JsonPropertyName("quiz_id")]
  ulong? QuizId = null,

  [property: JsonPropertyName("anonymous_submissions")]
  bool? AnonymousSubmissions = null,

  [property: JsonPropertyName("discussion_topic")]
  DiscussionTopicModel? DiscussionTopic = null,

  [property: JsonPropertyName("freeze_on_copy")]
  bool? FreezeOnCopy = null,

  [property: JsonPropertyName("frozen")]
  bool? Frozen = null,

  [property: JsonPropertyName("frozen_attributes")]
  IEnumerable<string>? FrozenAttributes = null,

  [property: JsonPropertyName("submission")]
  SubmissionModel? Submission = null,

  [property: JsonPropertyName("use_rubric_for_grading")]
  bool? UseRubricForGrading = null,

  [property: JsonPropertyName("rubric_settings")]
  object? RubricSettings = null,

  [property: JsonPropertyName("rubric")]
  IEnumerable<CanvasRubricCriteria>? Rubric = null,

  [property: JsonPropertyName("assignment_visibility")]
  IEnumerable<ulong>? AssignmentVisibility = null,

  [property: JsonPropertyName("overrides")]
  IEnumerable<CanvasAssignmentOverride>? Overrides = null,

  [property: JsonPropertyName("omit_from_final_grade")]
  bool? OmitFromFinalGrade = null,

  [property: JsonPropertyName("final_grader_id")]
  ulong? FinalGraderId = null,

  [property: JsonPropertyName("grader_comments_visible_to_graders")]
  bool? GraderCommentsVisibleToGraders = null,

  [property: JsonPropertyName("graders_anonymous_to_graders")]
  bool? GradersAnonymousToGraders = null,

  [property: JsonPropertyName("grader_names_anonymous_to_final_grader")]
  bool? GraderNamesVisibleToFinalGrader = null,

  [property: JsonPropertyName("anonymous_grading")]
  bool? AnonymousGrading = null
);

File: nextjs/src/models/canvas/assignments/canvasAssignment.ts
import { DiscussionTopicModel } from "../discussions/canvasDiscussionModelTopic";
import { SubmissionModel } from "../submissions/canvasSubmissionModel";
import { CanvasAssignmentDate } from "./canvasAssignmentDate";
import { CanvasAssignmentOverride } from "./canvasAssignmentOverride";
import { CanvasExternalToolTagAttributes } from "./canvasExternalToolTagAttributes";
import { CanvasLockInfo } from "./CanvasLockInfo";
import { CanvasRubricCriteria } from "./canvasRubricCriteria";
import { CanvasTurnitinSettings } from "./canvasTurnitinSettings";

export interface CanvasAssignment {
  id: number;
  name: string;
  description: string;
  created_at: string; // ISO 8601 date string
  has_overrides: boolean;
  course_id: number;
  html_url: string;
  submissions_download_url: string;
  assignment_group_id: number;
  due_date_required: boolean;
  max_name_length: number;
  peer_reviews: boolean;
  automatic_peer_reviews: boolean;
  position: number;
  grading_type: string;
  published: boolean;
  unpublishable: boolean;
  only_visible_to_overrides: boolean;
  locked_for_user: boolean;
  moderated_grading: boolean;
  grader_count: number;
  allowed_attempts: number;
  is_quiz_assignment: boolean;
  submission_types: string[];
  updated_at?: string; // ISO 8601 date string
  due_at?: string; // ISO 8601 date string
  lock_at?: string; // ISO 8601 date string
  unlock_at?: string; // ISO 8601 date string
  all_dates?: CanvasAssignmentDate[];
  allowed_extensions?: string[];
  turnitin_enabled?: boolean;
  vericite_enabled?: boolean;
  turnitin_settings?: CanvasTurnitinSettings;
  grade_group_students_individually?: boolean;
  external_tool_tag_attributes?: CanvasExternalToolTagAttributes;
  peer_review_count?: number;
  peer_reviews_assign_at?: string; // ISO 8601 date string
  intra_group_peer_reviews?: boolean;
  group_category_id?: number;
  needs_grading_count?: number;
  needs_grading_count_by_section?: {
    section_id: string;
    needs_grading_count: number;
  }[];
  post_to_sis?: boolean;
  integration_id?: string;
  integration_data?: any;
  muted?: boolean;
  points_possible?: number;
  has_submitted_submissions?: boolean;
  grading_standard_id?: number;
  lock_info?: CanvasLockInfo;
  lock_explanation?: string;
  quiz_id?: number;
  anonymous_submissions?: boolean;
  discussion_topic?: DiscussionTopicModel;
  freeze_on_copy?: boolean;
  frozen?: boolean;
  frozen_attributes?: string[];
  submission?: SubmissionModel;
  use_rubric_for_grading?: boolean;
  rubric_settings?: any;
  rubric?: CanvasRubricCriteria[];
  assignment_visibility?: number[];
  overrides?: CanvasAssignmentOverride[];
  omit_from_final_grade?: boolean;
  final_grader_id?: number;
  grader_comments_visible_to_graders?: boolean;
  graders_anonymous_to_graders?: boolean;
  grader_names_visible_to_final_grader?: boolean;
  anonymous_grading?: boolean;
}
